### ISSUE TITLE
Conditionally make assets...

### ISSUE THREAD

**Issue by jmervine** [heroku/docker-ruby#4]
Not 100% sure this will work but I think it's...

``` Dockerfile
# How to conditionally `rake assets:precompile`?
ONBUILD RUN bundle exec rake assets:precompile || true
```

**Comment by jmervine** [comment]
That's pretty hacky though.

**Comment by hone** [comment]
Yeah, if it fails it should break building if you expect it to run.

**Comment by jmervine** [comment]
Right, so you need to grep the output and return true if there's no task and false if there is... Ugly.

**Comment by gwmoura** [comment]
guys, I test this: 

``` bash
ONBUILD RUN if [ -f /app/user/Rakefile ]; then bundle exec rake assets:precompile; fi
```

and work for me. I tested using an simple script with Gemfile without rails or sinatra, only ruby

**Comment by jmervine** [comment]
@gwmoura I can see quite a few use-cases were an app would have a `Rakefile` without `assets:precompile`... how about...

```
ONBUILD RUN test "$(bundle exec rake -T | grep assets:precompile)" && bundle exec rake assets:precompile
```

**Comment by gwmoura** [comment]
@jmervine yes, can be, in my case, I execute a ruby script only, no Rails. Your solution not work for me :(

**Comment by jmervine** [comment]
@gwmoura Merging the two, you could do something like (untested):

```
ONBULD RUN test -f /app/user/Rakefile && (test "$(bundle exec rake -T | grep assets:precompile)" && bundle exec rake assets:precompile)
```

**Comment by gwmoura** [comment]
@jmervine I tested your script and this worked fine for me. @hone if you want, you can add on Dockerfile.

I tested locally, but I will add in my image based on `heroku/ruby`

Look my results:

``` bash
$ ls
app.rb  docker-compose.yml  Dockerfile  Gemfile  Gemfile.lock  run.sh  vendor

$ test -f Rakefile && (test "$(bundle exec rake -T | grep assets:precompile)" && bundle exec rake assets:precompile)
```

And in a rails app too not work

``` bash
$ ls
app  config     db       Gemfile.lock  log     Rakefile     test  vendor
bin  config.ru  Gemfile  lib           public  README.rdoc  tmp

$ test -f Rakefile && (test "$(bundle exec rake -T | grep assets:precompile)" && bundle exec rake assets:precompile)
I, [2015-10-01T18:41:24.746492 #2063]  INFO -- : Writing /home/george/workspace/tests/ruby/railsblocks/myproject/public/assets/application-d4d603d0ff049a3ecfbf019b468bf1a62a50b9fdafe41213c300fa071342b600.js
I, [2015-10-01T18:41:24.749187 #2063]  INFO -- : Writing /home/george/workspace/tests/ruby/railsblocks/myproject/public/assets/application-e80e8f2318043e8af94dddc2adad5a4f09739a8ebb323b3ab31cd71d45fd9113.css
```

**Comment by 3den** [comment]
why cant we just remove the assets:precompile from this Docker file and have users add it if needed?

**Comment by 3den** [comment]
even with the suggested fix it would not work for my team because we need to run `bower install` before the `assets:precompile`.

**Comment by jmervine** [comment]
@3den Perhaps add `bower install` as a pre-req task to `assets:precompile`?

**Comment by 3den** [comment]
@jmervine thanks for the reply, what is the best way to do that?

**Comment by jmervine** [comment]
@3den My rake-fu is not super strong, but the [first section here](http://jasonseifer.com/2010/04/06/rake-tutorial) talks about how to build dependency tasks. You'll have to build a custom task for the `bower install` part, I found an [overly complicated example here](https://github.com/rharriso/bower-rails/blob/master/lib/tasks/bower.rake), but you should be able to pull out the useful bits.

Because `assets:precompile` is built in to rails, or so I believe, you might have to do some googling around how to customize that task. It's probably not that difficult though.

Cheers!

Edit: [This talks about how to do exactly what we're talking about here.](http://stackoverflow.com/questions/27761643/how-can-i-run-custom-code-before-any-asset-precompile-in-rails) I didn't vet it, but it should be a good start.
